### PR TITLE
[CDF-21799] 🐉 modules upgrade command

### PR DESCRIPTION
**Pull request by doctrino**
# Description

Introduces a `cdf-tk modules upgrade my_project` command to automatically upgrade modules.

Note that this is a large PR, but the core of it is ~ 200 lines

A few comments
* Did not do tests as all, but instead have a script `python tests_migrations/run_module_upgrade.py` that runs the `modules upgrade against all previous versions. This makes it easier to debug, as well as setting up the required test data. The motivation for not having this as part of the regular test suite is that it would mean ~25MB extra for each past version we support. It is also a long
* The new `python tests_migrations/run_module_upgrade.py` replaces all the earlier migration tests, which can be removed when `modules upgraded` is part of the core package.
* Introduce the concept `Changed` to track changes, and added support back to `0.1.0`. This accounts for half of the code lines in this PR.
* Found an accidental bug in deploy which I wixed.

Some screenshots

image

image

image

image

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: CHANGELOG.cdf-tk.md [hunk 1]
```diff
@@ -17,13 +17,19 @@ Changes are grouped as follows:
...
 - When running `cdf-tk auth verify`, if the client does not have access to the `CDF_PROJECT` the user will now get
   a more informative error message.
 - When running `cdf-tk auth verify` and missing the `FunctionAcl(READ)` capability, the user will now get a more
   informative error message when checking the function service status
 
+### Added
+
+- Preview feature `MODULES_CMD` to allow interactive init and automatic upgrade of modules. Activate by running
+  `cdf-tk features set MODULES_CMD --enable`, and deactivate by running `cdf-tk features set MODULES_CMD --disable`.
+  Run `cdf-tk modules init/upgrade` to interactively initialize or upgrade modules.
```
Comment (doctrino) [inline CHANGELOG.cdf-tk.md:31]: Up for discussion, but i think we should log preview features in the changelog. It will be useful to know when they are available from. In addition, that enables us to do patch releases even if there have been no bugfixes for new features.

File: cognite_toolkit/_cdf_tk/prototypes/commands/modules.py [hunk 2]
```diff
@@ -245,3 +261,144 @@ def init(self, init_dir: Optional[str] = None, arg_package: Optional[str] = None
                 )
 
         raise typer.Exit()
+
+    def upgrade(self, project_dir: str | Path | None = None, verbose: bool = False) -> Changes:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/commands/modules.py:265]: This is the core change

File: cognite_toolkit/_cdf_tk/commands/deploy.py [hunk 3]
```diff
@@ -306,12 +306,23 @@ def to_create_changed_unchanged_triple(
...
+                are_equal = cdf_resource and loader.are_equal(item, cdf_resource)
+            except CogniteAPIError as e:
+                self.warn(
+                    MediumSeverityWarning(
+                        f"Failed to compare {loader.display_name} {loader.get_id(item)} for equality. Proceeding assuming not data in CDF. Error {e}."
+                    )
+                )
+                print(Panel(traceback.format_exc()))
+                are_equal = False
+            if are_equal:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/commands/deploy.py:319]: Accidental bug discovery. I was blocked by this so I solved it, and it was so small that I did not pull it out in its own PR.

**Comment by ronpal** [comment]
"False positive"?

Screenshot 2024-06-17 at 10 55 34

**Comment by ronpal** [comment]
Getting this error in run_module_update.py

Screenshot 2024-06-17 at 12 39 15

**Review by ronpal — APPROVED** [review]
